Can't extend a Java class in JRuby when the parent class is not in the system class loader

I'm starting my JVM process as JRuby `bundle console`, and then loading a bunch of Java stuff that uses non-system class loaders (technically it's Apache Karaf with Eclipse Equinox OSGi implementation). I then define a new Ruby class that inherits from a Java class in a Bundle that's not on the system class loader. I've added a loader with `JRuby.runtime.instance_config.add_loader` (essentially `org.jruby.embed.osgi.OSGiIsolatedScriptingContainer$BundleGetResources`, but I had to re-implement in Ruby since the former can't be used because org.osgi.framework.Bundle itself can't be found by the system class loader). This allows me to access the classes in the bundle from Ruby. But when I try to inherit from one, and it gets reified, I get `org/jruby/java/addons/ClassJavaAddons.java:65:in 'become_java!': requested class OpenHAB::DSL::Rules::AutomationRule was not reifiable (TypeError)`. Enabling `jruby.reify.logErrors=true`, I get additional logging:

```
2022-08-24T16:02:45.213-06:00 [main] ERROR RubyClass : failed to reify class OpenHAB::DSL::Rules::AutomationRule due to: 
java.lang.NoClassDefFoundError: org/openhab/core/automation/module/script/rulesupport/shared/simple/SimpleRule
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at org.jruby.dist/org.jruby.util.ClassDefiningJRubyClassLoader.defineClass(ClassDefiningJRubyClassLoader.java:53)
	at org.jruby.dist/org.jruby.RubyClass.reify(RubyClass.java:1368)
	at org.jruby.dist/org.jruby.RubyClass.reifyWithAncestors(RubyClass.java:1305)
	at org.jruby.dist/org.jruby.java.addons.ClassJavaAddons.becomeJava(ClassJavaAddons.java:91)
	at org.jruby.dist/org.jruby.java.addons.ClassJavaAddons.become_java(ClassJavaAddons.java:85)
	at org.jruby.dist/org.jruby.java.addons.ClassJavaAddons$INVOKER$s$become_java.call(ClassJavaAddons$INVOKER$s$become_java.gen)
	at org.jruby.dist/org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:824)
	at org.jruby.dist/org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:218)
	at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:372)
	at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:175)
...
Caused by: java.lang.ClassNotFoundException: org.openhab.core.automation.module.script.rulesupport.shared.simple.SimpleRule
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:476)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 233 more
```

(I wish I could see the 233 more, but I don't think it's critical to understand the problem. Digging into the JRuby code, there are a couple possibilities for the problem in `RubyClass.reify` (relevant code pasted for reference):
```java
        final byte[] classBytes = reifier.reify();

        final ClassDefiningClassLoader parentCL;
        if (parentReified.getClassLoader() instanceof OneShotClassLoader) {
            parentCL = (OneShotClassLoader) parentReified.getClassLoader();
        } else {
            if (useChildLoader) {
                parentCL = new OneShotClassLoader(runtime.getJRubyClassLoader());
            } else {
                parentCL = runtime.getJRubyClassLoader();
            }
        }
        boolean nearEnd = false;
        // Attempt to load the name we plan to use; skip reification if it exists already (see #1229).
        try {
            Class result = parentCL.defineClass(javaName, classBytes);
```
 * First, runtime.getJRubyClassLoader (which is returning a plain `org.jruby.util.JRubyClassLoader`) doesn't know to search the additional loaders I configured
 * Second, perhaps `parentCL` should actually be a `MultiClassLoader`, as used in `RealClassGenerator.defineRealImplClass`, which adds the superclass's loader as the parent of the `MultiClassLoader`, and then adds all of the `getExtraLoaders` individually, _and then_ wraps it in a `ClassDefiningJRubyClassLoader`.

I'm not familiar enough with JRuby internals to know if either or both of those approaches would have additional pitfalls.

**Environment Information**
 * `jruby 9.3.7.0 (2.6.8) 2022-08-16 c79ef237e0 OpenJDK 64-Bit Server VM 11.0.16+8-post-Ubuntu-0ubuntu122.04 on 11.0.16+8-post-Ubuntu-0ubuntu122.04 [x86_64-linux]`
 * `Linux opennuc 5.15.0-46-generic #49-Ubuntu SMP Thu Aug 4 18:03:25 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

(btw, thanks for https://blog.jruby.org/2022/06/jruby-class-extension-9.3 - while it didn't explain exactly my problem, it helped me understand the context more).

--- a/core/src/main/java/org/jruby/RubyClass.java
+++ b/core/src/main/java/org/jruby/RubyClass.java
@@ -60,6 +60,7 @@ import org.jruby.anno.JRubyMethod;
 import org.jruby.compiler.impl.SkinnyMethodAdapter;
 import org.jruby.exceptions.RaiseException;
 import org.jruby.internal.runtime.methods.DynamicMethod;
+import org.jruby.java.codegen.MultiClassLoader;
 import org.jruby.java.codegen.RealClassGenerator;
 import org.jruby.java.codegen.Reified;
 import org.jruby.java.proxies.ConcreteJavaProxy;
@@ -90,6 +91,7 @@ import org.jruby.util.ArraySupport;
 import org.jruby.util.ClassDefiningClassLoader;
 import org.jruby.util.CodegenUtils;
 import org.jruby.util.JavaNameMangler;
+import org.jruby.util.Loader;
 import org.jruby.util.OneShotClassLoader;
 import org.jruby.util.StringSupport;
 import org.jruby.util.collections.ConcurrentWeakHashMap;
@@ -1357,7 +1359,11 @@ public class RubyClass extends RubyModule {
             parentCL = (OneShotClassLoader) parentReified.getClassLoader();
         } else {
             if (useChildLoader) {
-                parentCL = new OneShotClassLoader(runtime.getJRubyClassLoader());
+                MultiClassLoader parentLoader = new MultiClassLoader(runtime.getJRubyClassLoader());
+                for(Loader cLoader : runtime.getInstanceConfig().getExtraLoaders()) {
+                    parentLoader.addClassLoader(cLoader.getClassLoader());
+                }
+                parentCL = new OneShotClassLoader(parentLoader);
             } else {
                 parentCL = runtime.getJRubyClassLoader();
             }

--- a/rakelib/test.rake
+++ b/rakelib/test.rake
@@ -25,11 +25,21 @@ namespace :test do
   desc "Compile test code"
   task :compile do
     mkdir_p "test/target/test-classes"
+    mkdir_p "test/target/test-classes-isolated/java_integration/fixtures/isolated"
+    mkdir_p "test/target/test-interfaces-isolated/java_integration/fixtures/isolated"
+
     classpath = %w[lib/jruby.jar test/target/junit.jar test/target/annotation-api.jar].join(File::PATH_SEPARATOR)
     # try detecting javac - so we use the same Java versions as we're running (JAVA_HOME) with :
     java_home = [ ENV_JAVA['java.home'], File.join(ENV_JAVA['java.home'], '..') ] # in case of jdk/jre
     javac = java_home.map { |home| File.expand_path('bin/javac', home) }.find { |javac| File.exist?(javac) } || 'javac'
     sh "#{javac} -cp #{classpath} -d test/target/test-classes #{Dir['spec/java_integration/fixtures/**/*.java'].to_a.join(' ')}"
+    # move the objects that need to be in separate class loaders
+    mv "test/target/test-classes/java_integration/fixtures/isolated/classes",
+       "test/target/test-classes-isolated/java_integration/fixtures/isolated",
+       force: true
+    mv "test/target/test-classes/java_integration/fixtures/isolated/interfaces",
+       "test/target/test-interfaces-isolated/java_integration/fixtures/isolated",
+       force: true
   end
 
   short_tests = ['jruby', 'mri']

--- a/spec/java_integration/fixtures/isolated/classes/GH7327Base.java
+++ b/spec/java_integration/fixtures/isolated/classes/GH7327Base.java
@@ -1,0 +1,4 @@
+package java_integration.fixtures.isolated.classes;
+
+public class GH7327Base {
+}

--- a/spec/java_integration/fixtures/isolated/interfaces/GH7327Interface.java
+++ b/spec/java_integration/fixtures/isolated/interfaces/GH7327Interface.java
@@ -1,0 +1,4 @@
+package java_integration.fixtures.isolated.interfaces;
+
+public interface GH7327Interface {
+}

--- a/spec/java_integration/reify/become_java_spec.rb
+++ b/spec/java_integration/reify/become_java_spec.rb
@@ -341,6 +341,37 @@ describe "JRuby class reification" do
     expect(klass.load_class(a_class.get_name)).to eq(a_class)
   end
 
+  it "supports reifying concrete classes extending classes from an unrelated class loader" do
+    isolated_class_path = File.expand_path('../../../../test/target/test-classes-isolated', __FILE__)
+    classloader = java.net.URLClassLoader.new([java.net.URL.new("file://#{isolated_class_path}/")].to_java(java.net.URL))
+    JRuby.runtime.instance_config.add_loader(classloader)
+
+    # we can reference it from JRuby, because it was added to the JRuby loader above
+    class GH7327 < Java::Java_integrationFixturesIsolatedClasses::GH7327Base; end
+    # it should have created a unique Java class
+    expect(GH7327.become_java!).not_to eq(Java::Java_integrationFixturesIsolatedClasses::GH7327Base.java_class)
+  ensure
+    JRuby.runtime.instance_config.extra_loaders.clear
+  end
+
+  it "supports reifying concrete classes extending classes and implementing interfaces from multiple unrelated class loaders" do
+    isolated_class_path = File.expand_path('../../../../test/target/test-classes-isolated', __FILE__)
+    classloader1 = java.net.URLClassLoader.new([java.net.URL.new("file://#{isolated_class_path}/")].to_java(java.net.URL))
+    JRuby.runtime.instance_config.add_loader(classloader1)
+    isolated_interface_path = File.expand_path('../../../../test/target/test-interfaces-isolated', __FILE__)
+    classloader2 = java.net.URLClassLoader.new([java.net.URL.new("file://#{isolated_interface_path}/")].to_java(java.net.URL))
+    JRuby.runtime.instance_config.add_loader(classloader2)
+
+    # we can reference them from JRuby, because they were added to the JRuby loader above
+    class GH7327WithInterface < Java::Java_integrationFixturesIsolatedClasses::GH7327Base
+      include Java::Java_integrationFixturesIsolatedInterfaces::GH7327Interface
+    end
+    # it should have created a unique Java class
+    expect(GH7327WithInterface.become_java!).not_to eq(Java::Java_integrationFixturesIsolatedClasses::GH7327Base.java_class)
+  ensure
+    JRuby.runtime.instance_config.extra_loaders.clear
+  end
+
   class ReifiedSample
     def hello; 'Sayonara from Ruby' end
     java_signature "java.lang.String ahoy()"

